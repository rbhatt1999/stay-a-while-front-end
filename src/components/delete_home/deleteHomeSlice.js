import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

const baseUrl = 'http://127.0.0.1:3000/home_stays';

export const getHomeStays = createAsyncThunk('home/getHomeStays', async (thunkAPI) => {
  const getURL = `${baseUrl}?my_home_stays=true`;
  const token = localStorage.getItem('token');
  const requestContent = {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${token}`,
    },
  };
  try {
    const respone = await axios.get(getURL, requestContent);
    return respone.data;
  } catch (error) {
    return thunkAPI.rejectWithValue(error.response.data.error);
  }
});

export const deleteHomeId = createAsyncThunk('home/deleteHomeId', async (homeStayId, thunkAPI) => {
  const deleteURL = `${baseUrl}/${homeStayId}`;
  const token = localStorage.getItem('token');
  const requestContent = {
    method: 'DELETE',
    headers: {
      Authorization: `Bearer ${token}`,
    },
  };
  try {
    await axios.delete(deleteURL, requestContent);
    return homeStayId;
  } catch (error) {
    return thunkAPI.rejectWithValue(error.response.data.error);
  }
});

const initialState = {
  loading: false,
  homeStayData: [],
  success: false,
  error: '',
  response: null,
};

const deleteHomeSlice = createSlice({
  name: 'homeStayData',
  initialState,
  reducers: { },
  extraReducers: (builder) => {
    builder
      .addCase(getHomeStays.pending, (state) => ({
        ...state,
        loading: true,
      }))
      .addCase(getHomeStays.fulfilled, (state, action) => ({
        ...state,
        loading: false,
        success: true,
        homeStayData: action.payload.data.home_stays,
      }))
      .addCase(getHomeStays.rejected, (state, action) => ({
        ...state,
        loading: false,
        success: false,
        error: action.payload,
      }))
      .addCase(deleteHomeId.pending, (state) => ({
        ...state,
        loading: true,
      }))
      .addCase(deleteHomeId.fulfilled, (state, action) => ({
        ...state,
        loading: false,
        success: true,
        homeStayData: state.homeStayData.filter((homeStay) => homeStay.id !== action.payload),
      }))
      .addCase(deleteHomeId.rejected, (state, action) => ({
        ...state,
        loading: false,
        success: false,
        error: action.payload,
      }));
  },
});

export default deleteHomeSlice.reducer;
