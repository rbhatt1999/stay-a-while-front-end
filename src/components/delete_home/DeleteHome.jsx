import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import ReactPaginate from 'react-paginate';
import { getHomeStays } from './deleteHomeSlice';
import DeleteCard from './deleteCard';
import style from './DeleteHome.module.css';

const DeleteHome = () => {
  const dispatch = useDispatch();
  const { homeStayData, loading } = useSelector((state) => state.deleteHome);
  const [currentPage, setCurrentPage] = useState(0);
  const [totalPages, setTotalPages] = useState(0);
  const itemsPerPage = 6;
  const startIndex = currentPage * itemsPerPage;
  const endIndex = startIndex + itemsPerPage;

  const handlePageChange = (selectedPage) => {
    setCurrentPage(selectedPage.selected);
  };

  useEffect(() => {
    dispatch(getHomeStays());
  },
  [dispatch]);

  useEffect(() => {
    if (homeStayData) {
      setTotalPages(Math.ceil(homeStayData.length / itemsPerPage));
    }
  }, [homeStayData]);

  const subset = homeStayData.slice(startIndex, endIndex);

  return (
    <div className={style.container}>
      <div className={style.cardContainer}>
        {loading && (
        <>
          <div className="loader align-self-center fixed-center-screen" />
        </>
        )}
        {homeStayData.length > 0 ? (
          subset.map((item) => (
            <DeleteCard item={item} key={item.id} />
          ))
        ) : (
          <h1>No homes to display</h1>
        )}
      </div>
      <ReactPaginate
        itemClass="pageitem"
        linkClass="pagelink"
        previousLabel="Prev"
        nextLabel="Next"
        pageCount={totalPages}
        onPageChange={handlePageChange}
        containerClassName="pagination justify-content-center"
        pageClassName="pageitem"
        pageLinkClassName="pagelink"
        activeClassName="active"
        previousClassName="pageitem"
        previousLinkClassName="pagelink"
        nextClassName="pageitem"
        nextLinkClassName="pagelink"
      />
    </div>
  );
};

export default DeleteHome;
