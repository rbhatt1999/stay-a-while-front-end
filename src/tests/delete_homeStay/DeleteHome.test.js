import React from 'react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { render, act, screen } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import axios from 'axios';
import deleteHomeSlice from '../../components/delete_home/deleteHomeSlice';
import DeleteHome from '../../components/delete_home/DeleteHome';
import homestays from './home_stays.json';

jest.mock('axios');

const store = configureStore({
  reducer: {
    deleteHome: deleteHomeSlice,
  },
});

describe('DeleteHome', () => {
  beforeEach(async () => {
    axios.get.mockResolvedValue({
      data: {
        data: {
          home_stays: [homestays.data.home_stays[0]],
        },
      },
    });

    await act(async () => {
      render(
        <Provider store={store}>
          <DeleteHome />
        </Provider>,
      );
    });
  });

  test('should have a submit button', () => {
    expect(screen.getByRole('button', { name: /Delete Home/i })).toBeInTheDocument();
  });

  test('should match screenshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <DeleteHome />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it should have an image of home stay', () => {
    expect(screen.getByRole('img')).toBeInTheDocument();
  });

  test('it should have a name of home stay', async () => {
    await act(() => render(
      <Provider store={store}>
        <DeleteHome />
      </Provider>,
    ));
    const homeName = await screen.findAllByText(homestays.data.home_stays[0].name);
    expect(homeName[0]).toBeInTheDocument();
  });

  test('it should have a location of home stay', async () => {
    await act(() => render(
      <Provider store={store}>
        <DeleteHome />
      </Provider>,
    ));
    const homeLocation = await screen.findAllByText(homestays.data.home_stays[0].location);
    expect(homeLocation[0]).toBeInTheDocument();
  });
});
