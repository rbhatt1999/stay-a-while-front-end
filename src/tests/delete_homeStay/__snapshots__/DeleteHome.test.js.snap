// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteHome should match screenshot 1`] = `
<div
  className="container"
>
  <div
    className="cardContainer"
  >
    <div
      className="card align-self-start deleteCard"
      style={
        Object {
          "width": "18rem",
        }
      }
    >
      <img
        alt="..."
        className="card-img-top cardImage"
        src="https://images.pexels.com/photos/2102587/pexels-photo-2102587.jpeg?auto=compress&cs=tinysrgb&w=600"
      />
      <div
        className="card-body"
      >
        <h5
          className="card-title"
        >
          Cottage by Maxwell
        </h5>
        <p
          className="card-text"
        >
          Sao Paulo, Brazil
        </p>
        <button
          className="btn btn-danger"
          onClick={[Function]}
          type="button"
        >
          Delete Home
        </button>
      </div>
    </div>
  </div>
  <ul
    aria-label="Pagination"
    className="pagination justify-content-center"
    role="navigation"
  >
    <li
      className="pageitem disabled"
    >
      <a
        aria-disabled="true"
        aria-label="Previous page"
        className="pagelink "
        onClick={[Function]}
        onKeyPress={[Function]}
        rel="prev"
        role="button"
        tabIndex="-1"
      >
        Prev
      </a>
    </li>
    <li
      className="pageitem"
    >
      <a
        aria-disabled="false"
        aria-label="Next page"
        className="pagelink"
        onClick={[Function]}
        onKeyPress={[Function]}
        rel="next"
        role="button"
        tabIndex="0"
      >
        Next
      </a>
    </li>
  </ul>
</div>
`;
